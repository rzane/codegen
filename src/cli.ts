import { program } from "commander";
import { generateAndWrite } from "./index";

const pkg = require("../package.json");

program
  .name("codegen")
  .version(pkg.version)
  .description("generate type definitions from GraphQL queries")
  .requiredOption(
    "-i, --input <input>",
    "glob of queries to generate types for"
  )
  .requiredOption(
    "-o, --output <output>",
    "path to file that will be generated"
  )
  .requiredOption(
    "-s, --schema <schema>",
    "URL or file path to a GraphQL schema"
  )
  .option("--suffix", "append suffix (e.g. Mutation, Query)")
  .option("--immutable", "generate readonly types")
  .action((opts) => {
    return generateAndWrite({
      input: opts.input,
      output: opts.output,
      schema: opts.schema,
      immutable: opts.immutable,
      suffix: opts.suffix,
    });
  })
  .parseAsync(process.argv);
