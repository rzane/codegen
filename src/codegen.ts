import { promises as fs } from "fs";
import { dirname } from "path";
import { parse, printSchema } from "graphql";
import { codegen } from "@graphql-codegen/core";
import { loadSchema, loadDocuments } from "@graphql-toolkit/core";
import { UrlLoader } from "@graphql-toolkit/url-loader";
import { JsonFileLoader } from "@graphql-toolkit/json-file-loader";
import { GraphQLFileLoader } from "@graphql-toolkit/graphql-file-loader";
import * as typescript from "@graphql-codegen/typescript";
import * as typescriptOperations from "@graphql-codegen/typescript-operations";
import * as typescriptReactApollo from "@graphql-codegen/typescript-react-apollo";

export interface GenerateConfig {
  input: string;
  output: string;
  schema: string;
  suffix?: boolean;
  immutable?: boolean;
}

const makeDefault = <T>(value: T | undefined, defaultValue: T): T => {
  return typeof value === "undefined" ? defaultValue : value;
};

/**
 * Generate code and write it to disk.
 */
export const generateAndWrite = async (
  config: GenerateConfig
): Promise<void> => {
  const code = await generate(config);
  const parent = dirname(config.output);

  await fs.mkdir(parent, { recursive: true });
  await fs.writeFile(config.output, code);
};

/**
 * Generate code and return it as a string.
 */
export const generate = async (config: GenerateConfig): Promise<string> => {
  const schemaAst = await loadSchema(config.schema, {
    loaders: [new UrlLoader(), new JsonFileLoader(), new GraphQLFileLoader()],
  });

  const documents = await loadDocuments(config.input, {
    loaders: [new GraphQLFileLoader()],
  });

  return codegen({
    filename: config.output,
    schema: parse(printSchema(schemaAst)),
    schemaAst,
    documents,
    plugins: [
      { typescript: {} },
      { typescriptOperations: {} },
      { typescriptReactApollo: {} },
    ],
    pluginMap: {
      typescript,
      typescriptOperations,
      typescriptReactApollo,
    },
    config: {
      withComponent: false,
      withHOC: false,
      withHooks: true,
      noNamespaces: true,
      preResolveTypes: true,
      omitOperationSuffix: makeDefault(config.suffix, true),
      immutableTypes: makeDefault(config.immutable, false),
      scalars: {
        DateTime: "string",
        Date: "string",
        Time: "string",
        JSON: "{ [key: string]: any }",
      },
    },
  });
};
