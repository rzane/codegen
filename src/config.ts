import * as path from "path";
import type { Types } from "@graphql-codegen/plugin-helpers";

export interface Options {
  root: string;
  schema: string;
  client: string;
  suffix: boolean;
  immutable: boolean;
  colocate: string | undefined;
}

const HOOKS = {
  afterAllFileWrite: ["prettier --write"],
  afterOneFileWrite: [],
  afterStart: [],
  beforeAllFileWrite: [],
  beforeDone: [],
  beforeOneFileWrite: [],
  onError: [],
  onWatchTriggered: [],
};

const SCALARS = {
  DateTime: "string",
  Date: "string",
  Time: "string",
};

function configure(opts: Options, plugins: string[]): Types.ConfiguredOutput {
  const config: Types.PluginConfig = {};

  const typescript = plugins.includes("typescript");
  const operations = plugins.includes("typescript-operations");
  const apollo = plugins.includes("typescript-react-apollo");

  if (typescript || operations) {
    config.scalars = SCALARS;
    config.preResolveTypes = true;
    config.immutableTypes = opts.immutable;
  }

  if (operations) {
    config.omitOperationSuffix = !opts.suffix;
  }

  if (apollo) {
    config.reactApolloVersion = 3;
  }

  return { plugins, config };
}

function buildDefault(opts: Options): Types.Config {
  const output = path.join(opts.root, "index.ts");
  const schemaOutput = path.join(opts.root, "schema.graphql");
  const documents = path.join(opts.root, "**/!(schema).graphql");

  return {
    schema: opts.schema,
    documents,
    hooks: HOOKS,
    generates: {
      [schemaOutput]: { plugins: ["schema-ast"] },
      [output]: configure(opts, [
        "typescript",
        "typescript-operations",
        `typescript-${opts.client}`,
      ]),
    },
  };
}

function buildColocate(opts: Options): Types.Config {
  const types = path.join(opts.colocate!, "index.ts");
  const baseTypesPath = path.relative(opts.root, opts.colocate!);

  const schemaOutput = path.join(opts.colocate!, "schema.graphql");
  const documents = path.join(opts.root, "**/!(schema).graphql");

  return {
    schema: opts.schema,
    documents,
    hooks: HOOKS,
    generates: {
      [schemaOutput]: { plugins: ["schema-ast"] },
      [types]: configure(opts, ["typescript"]),
      [opts.root]: {
        preset: "near-operation-file",
        presetConfig: {
          baseTypesPath,
          extension: ".ts",
        },
        ...configure(opts, [
          "typescript-operations",
          `typescript-${opts.client}`,
        ]),
      },
    },
  };
}

export function build(opts: Options): Types.Config {
  if (opts.colocate) {
    return buildColocate(opts);
  } else {
    return buildDefault(opts);
  }
}
