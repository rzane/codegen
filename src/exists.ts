import { promises as fs } from "fs";

export async function exists(path: string): Promise<boolean> {
  try {
    await fs.stat(path);
    return true;
  } catch (error) {
    if (error.code === "ENOENT") {
      return false;
    }

    throw error;
  }
}
