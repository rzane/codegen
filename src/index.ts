import { parse, printSchema } from "graphql";
import { codegen } from "@graphql-codegen/core";
import { loadSchema, loadDocuments } from "@graphql-toolkit/core";
import { UrlLoader } from "@graphql-toolkit/url-loader";
import { JsonFileLoader } from "@graphql-toolkit/json-file-loader";
import { GraphQLFileLoader } from "@graphql-toolkit/graphql-file-loader";
import * as typescript from "@graphql-codegen/typescript";
import * as typescriptOperations from "@graphql-codegen/typescript-operations";
import * as typescriptReactApollo from "@graphql-codegen/typescript-react-apollo";

export interface Options {
  input: string;
  output: string;
  schema: string;
  suffix?: boolean;
  immutable?: boolean;
}

export interface Result {
  schema: string;
  code: string;
}

/**
 * Generate code and return it as a string.
 */
export const generate = async (opts: Options): Promise<Result> => {
  const schemaAst = await loadSchema(opts.schema, {
    loaders: [new UrlLoader(), new JsonFileLoader(), new GraphQLFileLoader()],
  });

  const documents = await loadDocuments(opts.input, {
    loaders: [new GraphQLFileLoader()],
  });

  const schema = printSchema(schemaAst);
  const { suffix = false, immutable = false } = opts;

  const code = await codegen({
    filename: opts.output,
    schema: parse(schema),
    schemaAst,
    documents,
    plugins: [
      { typescript: {} },
      { typescriptOperations: {} },
      { typescriptReactApollo: {} },
    ],
    pluginMap: {
      typescript,
      typescriptOperations,
      typescriptReactApollo,
    },
    config: {
      noNamespaces: true,
      preResolveTypes: true,
      reactApolloVersion: 3,
      omitOperationSuffix: !suffix,
      immutableTypes: immutable,
      scalars: {
        DateTime: "string",
        Date: "string",
        Time: "string",
        JSON: "{ [key: string]: any }",
      },
    },
  });

  return { schema, code };
};
