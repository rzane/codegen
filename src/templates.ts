export const clientTemplate = `export function execute<Data, Variables>(
  query: string,
  variables?: Variables
): () => Promise<Data> {
  return async () => {
    const headers = {
      "Content-Type": "application/json",
    };

    const body = JSON.stringify({
      query,
      variables,
    });

    const response = await fetch("/graphql", {
      method: "POST",
      headers,
      body,
    });

    const json = await response.json();
    const error = json.errors && json.errors[0];

    if (error) {
      throw new Error(error.message);
    }

    return json.data;
  };
}`;
