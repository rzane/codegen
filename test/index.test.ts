import { join } from "path";
import { tmpdir } from "os";
import { promises as fs } from "fs";
import { randomBytes } from "crypto";
import { generate, generateAndWrite, GenerateOptions } from "../src";

const tmp = (...path: string[]): string => {
  const rand = randomBytes(10).toString("hex");
  return join(tmpdir(), rand, ...path);
};

const fixture = (...path: string[]): string => {
  return join(__dirname, "fixtures", ...path);
};

const makeConfig = (opts: Partial<GenerateOptions> = {}): GenerateOptions => ({
  output: tmp("index.ts"),
  schema: fixture("schema.graphql"),
  input: fixture("queries.graphql"),
  ...opts,
});

test("generates code", async () => {
  const config = makeConfig();
  const code = await generate(config);
  expect(code).toMatchSnapshot();
});

test("generate code with immutable types", async () => {
  const config = makeConfig({ immutable: true });
  const code = await generate(config);
  expect(code).toContain("readonly id: Scalars['ID']");
});

test("generate code with suffix", async () => {
  const config = makeConfig({ immutable: true });
  const code = await generate(config);
  expect(code).toContain("type CreatePostMutation");
});

test("generates code and writes it to disk", async () => {
  const config = makeConfig();
  await generateAndWrite(config);

  const stat = await fs.stat(config.output);
  expect(stat.size).toEqual(3939);
});
