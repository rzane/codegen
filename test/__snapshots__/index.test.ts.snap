// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generates code 1`] = `
"import gql from 'graphql-tag';
import * as ApolloReactCommon from '@apollo/react-common';
import * as ApolloReactHooks from '@apollo/react-hooks';
export type Maybe<T> = T | null;
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
};

export type Mutation = {
   __typename?: 'Mutation';
  createPost?: Maybe<Post>;
};


export type MutationCreatePostArgs = {
  data: PostInput;
};

export type Post = {
   __typename?: 'Post';
  id: Scalars['ID'];
  title: Scalars['String'];
};

export type PostInput = {
  title: Scalars['String'];
};

export type Query = {
   __typename?: 'Query';
  post?: Maybe<Post>;
};


export type QueryPostArgs = {
  id: Scalars['ID'];
};

export type PostVariables = {
  id: Scalars['ID'];
};


export type Post = { __typename?: 'Query', post?: Maybe<{ __typename?: 'Post', id: string, title: string }> };

export type CreatePostVariables = {
  data: PostInput;
};


export type CreatePost = { __typename?: 'Mutation', createPost?: Maybe<{ __typename?: 'Post', id: string, title: string }> };


export const PostDocument = gql\`
    query Post($id: ID!) {
  post(id: $id) {
    id
    title
  }
}
    \`;

/**
 * __usePost__
 *
 * To run a query within a React component, call \`usePost\` and pass it any options that fit your needs.
 * When your component renders, \`usePost\` returns an object from Apollo Client that contains loading, error, and data properties
 * you can use to render your UI.
 *
 * @param baseOptions options that will be passed into the query, supported options are listed on: https://www.apollographql.com/docs/react/api/react-hooks/#options;
 *
 * @example
 * const { data, loading, error } = usePost({
 *   variables: {
 *      id: // value for 'id'
 *   },
 * });
 */
export function usePost(baseOptions?: ApolloReactHooks.QueryHookOptions<Post, PostVariables>) {
        return ApolloReactHooks.useQuery<Post, PostVariables>(PostDocument, baseOptions);
      }
export function usePostLazyQuery(baseOptions?: ApolloReactHooks.LazyQueryHookOptions<Post, PostVariables>) {
          return ApolloReactHooks.useLazyQuery<Post, PostVariables>(PostDocument, baseOptions);
        }
export type PostHookResult = ReturnType<typeof usePost>;
export type PostLazyQueryHookResult = ReturnType<typeof usePostLazyQuery>;
export type PostQueryResult = ApolloReactCommon.QueryResult<Post, PostVariables>;
export const CreatePostDocument = gql\`
    mutation CreatePost($data: PostInput!) {
  createPost(data: $data) {
    id
    title
  }
}
    \`;
export type CreatePostMutationFn = ApolloReactCommon.MutationFunction<CreatePost, CreatePostVariables>;

/**
 * __useCreatePost__
 *
 * To run a mutation, you first call \`useCreatePost\` within a React component and pass it any options that fit your needs.
 * When your component renders, \`useCreatePost\` returns a tuple that includes:
 * - A mutate function that you can call at any time to execute the mutation
 * - An object with fields that represent the current status of the mutation's execution
 *
 * @param baseOptions options that will be passed into the mutation, supported options are listed on: https://www.apollographql.com/docs/react/api/react-hooks/#options-2;
 *
 * @example
 * const [createPost, { data, loading, error }] = useCreatePost({
 *   variables: {
 *      data: // value for 'data'
 *   },
 * });
 */
export function useCreatePost(baseOptions?: ApolloReactHooks.MutationHookOptions<CreatePost, CreatePostVariables>) {
        return ApolloReactHooks.useMutation<CreatePost, CreatePostVariables>(CreatePostDocument, baseOptions);
      }
export type CreatePostHookResult = ReturnType<typeof useCreatePost>;
export type CreatePostMutationResult = ApolloReactCommon.MutationResult<CreatePost>;
export type CreatePostMutationOptions = ApolloReactCommon.BaseMutationOptions<CreatePost, CreatePostVariables>;"
`;
